### ISSUE TITLE
[TensorExpr] Loop transformations: IR Simplifier fails to fold index expression after loop flattening

### ISSUE THREAD

**Issue by ZolotukhinM** [pytorch/pytorch#51173]
## 🐛 Bug

If the original computation had 3 or more levels of loops and we flatten them into a single loop, then the index expression should become trivial. However, it doesn't get folded and we end up with a complicated expression there.

In other words, here is the sequence of transformations I'd like to be able to perform:
Original loop:
```
for i in 0..N:
  for j in 0..M:
    for k in 0..K:
      a[i,j,k] = 0
```
--->  flatten loops --->
```
for t in 0..N*M*K:
  a[t/M/K, t/K%M, t%K] = 0
```
---> flatten indices --->
```
for t in 0..N*M*K:
  a[(t/M/K)*(M*K) + (t/K%M)*K + t%K] = 0
```
---> simplify (**missing now**) --->
```
for t in 0..N*M*K:
  a[t] = 0
```

## To Reproduce
```
import torch

class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C.te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

with kernel_arena_scope():
    dtype = torch._C.te.Dtype.Float
    M = 10
    K = 9
    N = 7

    def get_dim_args(dims):
        dim_args = []
        for dim in dims:
            dim_args.append(torch._C.te.DimArg(dim, 'i' + str(len(dim_args))))
        return dim_args

    (MM, KK, NN) = [torch._C.te.ExprHandle.int(x) for x in [M, K, N]]

    dtype = torch._C.te.Dtype.Float
    X = torch._C.te.Placeholder('X', dtype, [MM, KK, NN])
    Y = torch._C.te.Compute('T', get_dim_args([MM, KK, NN]),
            lambda m, n, k: X.load([m, n, k]))
    loopnest = torch._C.te.LoopNest([Y])
    loops = loopnest.get_loops_for(Y)
    loopnest.flatten(loops)
    loopnest.prepare_for_codegen()
    loopnest.simplify()
    s = loopnest.root_stmt()
    print(s)
```
## Current behavior
The script above prints this:
```
{
  for (int i0_flat = 0; i0_flat < 630; i0_flat++) {
    T[(7 * ((i0_flat / 7) % 9) + i0_flat % 7) + 63 * (i0_flat / 63)] = X[(7 * ((i0_flat / 7) % 9) + i0_flat % 7) + 63 * (i0_flat / 63)];
  }
}
```

## Expected behavior
The script above prints this:
```
{
  for (int i0_flat = 0; i0_flat < 630; i0_flat++) {
    T[i0_flat] = X[i0_flat];
  }
}
```

**Comment by huiguoo** [comment]
I think we'll need to extend the flattened indices' pattern from 2 dimension to N dimensions. 

The pattern for 2 dimensions is,
dims  = (K<sub>2</sub>, K<sub>1</sub>), t in K<sub>2</sub> * K<sub>1</sub></code>      
flattened index = t/K<sub>1</sub>*K<sub>1</sub> + t%K<sub>1</sub>     


The pattern for N dimensions will be,
dims = (K<sub>n</sub>, K<sub>n-1</sub>, ... K<sub>3</sub>, K<sub>2</sub>, K<sub>1</sub>), t in K<sub>n</sub> * K<sub>n-1</sub> * ... * K<sub>1</sub>
flattened index = 
(t/(K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub> * K<sub>n-1</sub>)) * (K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub> * K<sub>n-1</sub>) + 
(t/(K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub>)%K<sub>n-1</sub>) * (K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub>) +
... +
(t/(K<sub>1</sub> * K<sub>2</sub>) %K<sub>3</sub>) * (K<sub>1</sub> * K<sub>2</sub>) +
(t/K<sub>1</sub> %K<sub>2</sub>) * K<sub>1</sub> +
t%K<sub>1</sub>

In fact, to be more clear, the first addend
(t/(K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub> * K<sub>n-1</sub>)) * (K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub> * K<sub>n-1</sub>)
equals to,
(t/(K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub> * K<sub>n-1</sub>)%K<sub>n</sub>) * (K<sub>1</sub> * K<sub>2</sub> * ... * K<sub>n-2</sub> * K<sub>n-1</sub>)

**Comment by ZolotukhinM** [comment]
Yep, that looks right! I think the current 2d pattern is currently represented via `RoundOff` expression - we might need to extend its semantics to cover these cases (but that seems like a good thing to do).

**Comment by huiguoo** [comment]
Fixed in PR #52882
